state/binarystorage/binarystorage.go

```diff
@@ -63,15 +64,15 @@ func New(
 func (s *binaryStorage) Add(ctx context.Context, r io.Reader, metadata Metadata) (resultErr error) {
 	// Add the binary file to storage.
 	path := fmt.Sprintf("tools/%s-%s", metadata.Version, metadata.SHA256)
-	if err := s.managedStorage.Put(context.TODO(), path, r, metadata.Size); err != nil {
+	if err := s.managedStorage.Put(ctx, path, r, metadata.Size); err != nil && !errors.Is(err, objectstoreerrors.ErrHashAlreadyExists) {
```

Comment:
This might be overly cautious, but maybe we should check the size/path is the same as the existing binary to be sure we're not overwriting something else?

Reply:
So, the data (what is contained in the io.Reader) is content addressable with the size not just on the path. But yes, we should switch over to the hash version of the call, so it ensures it matches the exact hash, not just the size. The object store already supports this, the managed tooling isn't using it.